SSA/Core/Framework.lean

```diff
@@ -220,282 +590,717 @@ args, since there now are equation lemmas for it.
-    -- Was just `simp only [addProgramToLets, ih, Com.denote]`
-    rw [addProgramToLets]
-    simp only [ih, Com.denote]
-    congr
-    funext t v
-    cases v using Var.casesOn
-    . rfl
-    . simp [Lets.denote]; rfl
-
-/-- Add some `Lets` to the beginning of a program -/
-def addLetsAtTop : (lets : Lets Op Γ₁ Γ₂) → (inputProg : Com Op Γ₂ t₂) → Com Op Γ₁ t₂
-  | Lets.nil, inputProg => inputProg
-  | Lets.lete body e, inputProg =>
-    addLetsAtTop body (.lete e inputProg)
-
-theorem denote_addLetsAtTop :
-    (lets : Lets Op Γ₁ Γ₂) → (inputProg : Com Op Γ₂ t₂) →
-    (addLetsAtTop lets inputProg).denote =
-      inputProg.denote ∘ lets.denote
-  | Lets.nil, inputProg => rfl
-  | Lets.lete body e, inputProg => by
-    rw [addLetsAtTop, denote_addLetsAtTop body]
-    funext
-    simp only [Com.denote, Ctxt.Valuation.snoc, Function.comp_apply, Lets.denote,
-      eq_rec_constant]
-    congr
-    funext t v
-    cases v using Var.casesOn <;> simp
-
-/-- `addProgramInMiddle v map lets rhs inputProg` appends the programs
-`lets`, `rhs` and `inputProg`, while reassigning `v`, a free variable in
-`inputProg`, to the output of `rhs`. It also assigns all free variables
-in `rhs` to variables available at the end of `lets` using `map`. -/
-def addProgramInMiddle {Γ₁ Γ₂ Γ₃ : Ctxt Ty} (v : Var Γ₂ t₁)
-    (map : Γ₃.Hom Γ₂)
-    (lets : Lets Op Γ₁ Γ₂) (rhs : Com Op Γ₃ t₁)
-    (inputProg : Com Op Γ₂ t₂) : Com Op Γ₁ t₂ :=
-  let r := addProgramToLets lets map rhs
-  addLetsAtTop r.lets <| inputProg.changeVars (r.diff.toHom.with v r.var)
-
-theorem denote_addProgramInMiddle {Γ₁ Γ₂ Γ₃ : Ctxt Ty}
-    (v : Var Γ₂ t₁) (s : Valuation Γ₁)
-    (map : Γ₃.Hom Γ₂)
-    (lets : Lets Op Γ₁ Γ₂) (rhs : Com Op Γ₃ t₁)
-    (inputProg : Com Op Γ₂ t₂) :
-    (addProgramInMiddle v map lets rhs inputProg).denote s =
-      inputProg.denote (fun t' v' =>
-        let s' := lets.denote s
-        if h : ∃ h : t₁ = t', h ▸ v = v'
-        then h.fst ▸ rhs.denote (fun t' v' => s' (map v'))
-        else s' v') := by
-  simp only [addProgramInMiddle, Ctxt.Hom.with, denote_addLetsAtTop, Function.comp_apply,
-              Com.denote_changeVars]
-  congr
-  funext t' v'
-  split_ifs
-  next h =>
-    rcases h with ⟨⟨⟩, ⟨⟩⟩
-    simp [denote_addProgramToLets_var]
-  next h₁ h₂ =>
-    rcases h₁ with ⟨⟨⟩, ⟨⟩⟩
-    simp at h₂
-  next h₁ h₂ =>
-    rcases h₂ with ⟨⟨⟩, ⟨⟩⟩
-    simp at h₁
-  next =>
-    apply denote_addProgramToLets_lets
-
-structure FlatCom (Op : _) {Ty : _} [OpSignature Op Ty] (Γ : Ctxt Ty) (t : Ty) where
-  {Γ_out : Ctxt Ty}
-  /-- The let bindings of the original program -/
-  lets : Lets Op Γ Γ_out
-  /-- The return variable -/
-  ret : Var Γ_out t
-
-def Com.toLets {t : Ty} : Com Op Γ t → FlatCom Op Γ t :=
-  go .nil
-where
-  go {Γ_out} (lets : Lets Op Γ Γ_out) : Com Op Γ_out t → FlatCom Op Γ t
-    | .ret v => ⟨lets, v⟩
-    | .lete e body => go (lets.lete e) body
 
+-- TODO: this doesn't morally fit here, but we can't yoink it up, figure out what to do
```

Comment:
IMO we should probably split this file up anyway (but that's maybe a question for later)

Reply:
Yes, let's do this post-merge, please!